src/os/bluestore/BlueStore.cc

```diff
@@ -12399,6 +12399,40 @@ void BlueStore::_zoned_clean_zone(uint64_t zone_num) {
   dout(10) << __func__ << " cleaning zone " << zone_num << dendl;
   // TODO: (1) copy live objects from zone_num to a new zone, (2) issue a RESET
   // ZONE operation to the device for the corresponding zone.
+
+  zone_state_t zone_state;
+  std::string pfx = _zoned_get_prefix(zone_num * bdev->get_zone_size());
+  KeyValueDB::Iterator it = db->get_iterator(pfx, KeyValueDB::ITERATOR_NOCACHE);
+
+  while (it->valid())
+  {
+    std::string k = it->key();
+    bufferlist bl = it->value();
+    auto p = bl.cbegin();
+    int64_t offset;
+    ::decode(offset, p);
+    dout(40) << __func__ << " Copying object with key (zone_num + oid): " << k << " and offset: " << offset << dendl;
+    ghobject_t oid;
+    int r = get_key_object(k, &oid);
+    CollectionRef c = _get_collection(coll_t::meta());
```

Comment:
one should use collection related to the object not 'meta' one here. Apparently you should persist that somehow - the issue is that collection might split/merge over time...

Reply:
@ifed01 Hmmm, so we should persist collection id as well. I'm being lazy here, but with the rados tool, I can write an object and then retrieve the object using just its name. So what am I missing here? Why can't we do the same using the low-level API?

Reply:
Onodes are addressable via a <cid, oid> pair, single oid isn't enough. And every onode is bound to a specific collection hence one needs to get proper collection reference via _get_collection(cid) call to access that onode first. 

The above _get_collection(coll_t::meta()) returns just a "meta" collection which keeps just a few particular onodes, not all of them. 
So one needs to preserve onode's cid when persisting onode information in _zoned_update_cleaning_metadata() and reuse that cid here...